pkg/storegateway/bucket_index_postings.go

```diff
@@ -322,3 +324,143 @@ func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
 	}
 	return p
 }
+
+type postingsSelectionStrategy interface {
+	// name should be a static string which identifies this strategy.
+	// The return value can be later used to find the strategy which yielded a partitioning of postingGroups.
+	name() string
+
+	// selectPostings can modify the passed slice of posting groups.
+	selectPostings([]postingGroup) (selected, omitted []postingGroup)
+}
+
+type selectAllStrategy struct{}
+
+func (selectAllStrategy) name() string {
+	return "all"
+}
+
+func (selectAllStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	return groups, nil
+}
+
+// worstCaseFetchedDataStrategy select a few of the posting groups such that their total size
+// does not exceed the size of series in the worst case. The worst case is fetching all series
+// in the smallest non-subtractive posting group - this is effectively the
+// upper bound on how many series all the posting groups can select after being intersected/subtracted.
+//
+// This strategy is meant to prevent fetching some the largest posting lists in an index. Those are usually
+// less selective the smaller ones and only add cost to fetching.
+//
+// The strategy greedily selects the first N posting groups (sorted in ascending order by their size)
+// whose combined size doesn't exceed the size of series in the worst case. The rest of the posting groups
+// are omitted.
+// worstCaseFetchedDataStrategy uses a fixed estimation about the size of series in the index (tsdb.EstimatedSeriesP99Size).
+type worstCaseFetchedDataStrategy struct {
+	// postingListActualSizeFactor affects how
+	// posting lists are summed together.
+	// This factor is multiplied by the size of posting lists
+	// in the block index.
```

Comment:
[nit] I would rather explain "why" as you did in the PR description, explaining that the actual postings list size different whether we load it from cache or looking up the index. Since we don't know the cache hit ratio before actually fetching it, the factor we use here is based on an estimation of the cache hit ratio.